pylint/config/caching.py

```diff
@@ -0,0 +1,69 @@
+# Licensed under the GPL: https://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+# For details: https://github.com/PyCQA/pylint/blob/main/LICENSE
+# Copyright (c) https://github.com/PyCQA/pylint/blob/main/CONTRIBUTORS.txt
+
+"""Caching functions.
+
+These are in pylint.config for compatibility reasons.
```

Comment:
Sure it make sense, especially considering I already did that for PYLINT_HOME with pylint.constants.

Reply:
Note that I'm not sure `pylint/lint` is the best place? Perhaps we want a `caching` module? I'm not sure..

Reply:
This create a circular import between lint/base_checker.py/config/pylinter.
```
ImportError while loading conftest '/home/pierre/pylint/tests/conftest.py'.
tests/conftest.py:14: in <module>
    from pylint import checkers
pylint/checkers/__init__.py:49: in <module>
    from pylint.checkers.base_checker import (
pylint/checkers/base_checker.py:17: in <module>
    from pylint.config.arguments_provider import _ArgumentsProvider
pylint/config/__init__.py:38: in <module>
    from pylint.lint.caching import load_results as _real_load_result
pylint/lint/__init__.py:21: in <module>
    from pylint.lint.pylinter import PyLinter
pylint/lint/pylinter.py:24: in <module>
    from pylint.checkers.base_checker import BaseChecker
E   ImportError: cannot import name 'BaseChecker' from partially initialized module 'pylint.checkers.base_checker' (most likely due to a circular import) (/home/pierre/pylint/pylint/checkers/base_checker.py)
```
I'd like to do it though 😄 Do you think we can load lazily for the ``pylint.config`` api with ``lazy_object_proxy`` (This is done a lot in astroid but for real circular import, not for deprecated API)  ?

Reply:
Does it also happen when you put `LinterStats` under `if TYPE_CHECKING`? This file doesn't really import anything? So if you import directly from it it shouldn't create any new circles I think?

Reply:
I was going to detail the circular import but it will be clearer if you check the stacktract. I don't think LinterStats is involved in the circular import. I think we might fix it if we lazily import load and save results in ``config.__init__``

Reply:
Not a big fan of lazy importing (it doesn't actually solve the circular import).
Could you push your current solution to a branch? I don't understand:
```python
pylint/config/__init__.py:38: in <module>
    from pylint.lint.caching import load_results as _real_load_result
pylint/lint/__init__.py:21: in <module>
    from pylint.lint.pylinter import PyLinter
```

Why is `lint/__init__.py` being imported there?

Reply:
https://github.com/PyCQA/pylint/pull/6468/commits/b15bb48c4bad8e5690d7c852242e32060c0face8

Reply:
Seems like it was fixed?

Reply:
Yeah, semi lazily was enough I don't know why I wanted to use ``lazy_object_proxy``